### DISCUSSION TITLE
Executable Books for Pluto

### DISCUSSION THREAD

**Discussion by jdc-pub** [fonsp/Pluto.jl#2106] [Pluto development]
Pluto is fantastic, I use it all the time, and lately I've been interested in using Pluto to write small blog posts with computational elements. I want to string a bunch of those posts into a "series", and eventually have a note-set on the topic I'm writing about. Ideally you could use Pluto's HTML export feature for rich web-viewing, and use [`PlutoStaticHTML.jl`](https://huijzer.xyz/PlutoStaticHTML.jl/dev/) to export each notebook to a more plain format for integration with a `Documenter.jl` site, and/or for including in a PDF format. 

I just found the [Executable Books](https://executablebooks.org/en/latest/) project, along with [Jupyter Books](https://jupyterbook.org/en/stable/intro.html). 

Are there plans in-works for developing a kind of "executable book" format with Pluto? I'd be interested in helping out with that project!

**Comment by rikhuijzer** [comment]
> Are there plans in-works for developing a kind of "executable book" format with Pluto? I'd be interested in helping out with that project!

100%. I was just talking with Fons about it last week and he gave me some great tips on how to do it. Basically, `PlutoStaticHTML.jl` can already go a long way. The only things that are needed as far as I can tell is

- Something to wrap figures in so that you can set a caption and a label. Just like in LaTeX.
- Some way to manage references. Or, use footnotes for references as a start could also work.
- Some way to specify in which order the notebooks should be put in the PDF
- Some way to specify LaTeX formatting

All these points a very much doable. I did it for [Books.jl](https://github.com/JuliaBooks/Books.jl) and it wasn't too hard.

> Ideally you could use Pluto's HTML export feature for rich web-viewing, and use [PlutoStaticHTML.jl](https://huijzer.xyz/PlutoStaticHTML.jl/dev/) to export each notebook to a more plain format for integration with a Documenter.jl site, and/or for including in a PDF format.

So a video in the HTML and an image in the PDF? I personally think it's easier to keep both outputs as similar as possible otherwise you have to verify two formats when developing. If they look very similar, you only have to check one. But, I'd be happy to be proven wrong.

Concretely, I'm more than happy to work with other people on this. In the next weeks, I won't have much time though. If you wanna start, then you could try to fix a small feature toward executable Pluto books. For example, figure out a way to set captions to images. Maybe by a package like `PlutoUI` or so.

**Reply by jdc-pub** [reply]
> So a video in the HTML and an image in the PDF? I personally think it's easier to keep both outputs as similar as possible otherwise you have to verify two formats when developing. If they look very similar, you only have to check one. But, I'd be happy to be proven wrong.

Oh I just meant that we could have export options for typical Pluto notebooks, a web-format (like `Documenter` or `Books`), and a PDF output. I hadn't thought about how rich media like videos would show in the PDF.

> Concretely, I'm more than happy to work with other people on this. In the next weeks, I won't have much time though. If you wanna start, then you could try to fix a small feature toward executable Pluto books. For example, figure out a way to set captions to images. Maybe by a package like PlutoUI or so.

I have a bit of time on weekends, though this summer will be pretty busy. I'll poke around `PlutoUI` and `Pluto` for ways to try to add captions.

**Reply by jdc-pub** [reply]
@rikhuijzer I'm starting to think tying into the [Executable Books](https://executablebooks.org/en/latest/) is worth consideration, as opposed to developing all of this functionality fresh. Julia provides tons of composability. If we had a `MyST` package, like how I describe [here](https://discourse.julialang.org/t/has-anyone-added-a-myst-markdown-extension-to-julia/83289), then we could tie into the features that the **Educational Books** folks have already developed: citations, figures, cross-referencing, and more. 

What do you think?

**Reply by rikhuijzer** [reply]
Glad to hear that you’re still thinking about this problem. From my experience, a pure Julia implementation is a better bet long term. Books.jl relied heavily on Pandoc which meant that some bugs could only be solved at Pandoc which took considerable efforts to figure out and work around.

**Reply by rikhuijzer** [reply]
I’ve started at PDF generation by the way and plan to tie PlutoStaticHTML into Books. Current work is at https://github.com/rikhuijzer/PlutoStaticHTML.jl/pull/124.

**Comment by stefanbringuier** [comment]
I'm curious about the state of this idea. Does [PlutoStaticHTML.jl](https://github.com/rikhuijzer/PlutoStaticHTML.jl) achieve this yet? Also, does PlutoStaticHTML.jl enable the cross-referencing between notebooks?  I'm not sure if it's possible to link directly to sections between notebooks with relative paths and common markdown in Pluto, but maybe just the notebook files, ex., `[Notebook_2](./open?path=Notebook_2.jl)`. Is it possible to specify the ordering of notebooks with PlutoStaticHTML.jl and generate section numbers? I'm asking since I think it would be useful to essentially create a book from collection Pluto notebooks.

**Reply by rikhuijzer** [reply]
PlutoStaticHTML.jl doesn't implement this via some kind of official API, no. For HTML, you could use regular links though.

**Reply by stefanbringuier** [reply]
I see. I did come across recent Pluto.jl [call notes ](https://juliapluto.github.io/weekly-call-notes/2023/05-16/notes.html) and there is mention of potential Quarto+Pluto support, which would awesome.
